### PR TITLE
Optimize floating mesh mapping into background triangulation

### PR DESCRIPTION
**Pull request by blaisb**
# Description of the problem

- The current DEM solver map every triangle of the floating meshes inside of the background triangulation for the particle-floating mesh contact detection. Although robust, this is very slow and not practical.

# Description of the solution

- I have added a new mechanism where the triangles are mapped in the background triangulation using the distance between the center of the cell and the triangles

# How Has This Been Tested?

- I tested it on all the examples and it works well
- A unit test has been added for the mapping
- An application test has been added with a slowly moving wall. I used this to debug also
- I took the liberty of improving the granular mixer case. It was far too slow.

# Documentation

[] I need to finalize the granular mixer documentation with the updates i did to the case.

# Future changes

- I will keep on optimizing this in the future, but this is a first work.

# Comments

- This already gives a considerable speed-up. The dambreak case on my machine went from 8.5 minutes to 5.5 minutes. The mixer case is even more intense.

### CODE REVIEW

File: source/core/lethe_grid_tools.cc
Comment (acdaigneault) [inline source/core/lethe_grid_tools.cc:1482]: I know it's a math thing and it doesn't really represent anything obvious, but some short explanation here and there would help :)

File: source/core/lethe_grid_tools.cc
Comment (acdaigneault) [inline source/core/lethe_grid_tools.cc:1376]: It's called `pos` here, but in the .h it's `point`

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:43]: CLion shows me a lot of grey include meaning some of them may be used in this class. Are all the current includes used ?

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:110]: Can you explain why?

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:115]:
```suggestion
  // Calculate distance from cell center to solid_cell
```

Comment (acdaigneault) [reply source/core/serial_solid.cc:115]: Cell center to solid cell face/surface ?

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:450]: Why those brackets? Why not just a short comment to separate the steps?

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:445]: extra line

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:465]: extra lines

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:511]: Extra line :p

File: source/core/serial_solid.cc
Comment (acdaigneault) [inline source/core/serial_solid.cc:570]: Why all those extra lines :(? Is there a purpose? Because there are some after this comment, but I will stop x)

File: source/dem/dem_container_manager.cc
Comment (acdaigneault) [inline source/dem/dem_container_manager.cc:396]: BB, do you still need this note?

File: source/dem/particle_wall_broad_search.cc
Comment (acdaigneault) [inline source/dem/particle_wall_broad_search.cc:163]: Same here

**Review by acdaigneault — CHANGES_REQUESTED** [review]
It seems like a difficult work here, that's nice! My major comment is that I think some of the new implementation needs more description and comments in the code since the variable names are not very descriptive.

File: include/core/serial_solid.h
Comment (blaisb) [reply include/core/serial_solid.h:131]: Agreed, but I am kinda trying to be systematic here :).

File: source/core/lethe_grid_tools.cc
Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1402]: Yeah these were taken from a reference directly. (Eberly 2003). These are just formulas, the variables don't really have a concrete name...

File: source/core/lethe_grid_tools.cc
Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1416]: Refers to the reference above that is commented (Eberly 2003)

File: source/core/lethe_grid_tools.cc
Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1482]: Yeah, I'll just add a more complete reference to the original document, it's kinda hard to explain this without rewriting the entire reference.

File: source/core/serial_solid.cc
Comment (blaisb) [reply source/core/serial_solid.cc:43]: Most likely not. I can't spot them from my editor regrettably,.

File: source/core/serial_solid.cc
Comment (blaisb) [reply source/core/serial_solid.cc:110]: Well not having static memory allocation is more expensive? I'm not sure what requires an explanation?

File: source/core/serial_solid.cc
Comment (blaisb) [reply source/core/serial_solid.cc]: Well should be triangle

File: source/core/serial_solid.cc
Comment (blaisb) [reply source/core/serial_solid.cc:450]: Because otherwise I need to have two sets of variable with different names (the names, the interpretation, etc.), I felt it was a bit easier this way

File: source/core/serial_solid.cc
Comment (blaisb) [reply source/core/serial_solid.cc:570]: Not sure really. I coded this over multiple weeks so things might have drifted.

File: source/dem/dem_container_manager.cc
Comment (blaisb) [reply source/dem/dem_container_manager.cc:396]: Nope

File: source/dem/particle_wall_broad_search.cc
Comment (blaisb) [reply source/dem/particle_wall_broad_search.cc:163]: This one will remain, I'm still skeptical of the comment below.

File: source/core/lethe_grid_tools.cc
Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1376]: It would theoretically work in 2D yes

File: source/core/lethe_grid_tools.cc
Comment (acdaigneault) [reply source/core/lethe_grid_tools.cc:1487]: Ok!

File: source/core/serial_solid.cc
Comment (acdaigneault) [reply source/core/serial_solid.cc:110]: Why do you need static memory, is it for the previous vertices of the mesh before displacement?

File: source/core/serial_solid.cc
Comment (acdaigneault) [reply source/core/serial_solid.cc:450]: Oh I didn't know we should do that, nice

File: source/core/lethe_grid_tools.cc
Comment (acdaigneault) [inline source/core/lethe_grid_tools.cc:1379]: Here I meant for the auto -> Point<dim>, but not important

Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1380]: Ohh no I can change that :)

File: tests/core/serial_solid_intersection_01.cc
Comment (voferreira) [inline tests/core/serial_solid_intersection_01.cc:86]:
```suggestion
  // Loop over background
```

**Review by voferreira — COMMENTED** [review]
There is basically just one question regarding region 4. The rest of the regions look good to me.

File: source/core/lethe_grid_tools.cc
Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1402]: Agreed, I'll put it there instead

File: source/core/lethe_grid_tools.cc
Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1423]: I'm pretty convinced it's d<0

Comment (blaisb) [reply source/core/lethe_grid_tools.cc:1423]: Look at this reference for more information
https://www.geometrictools.com/Documentation/DistancePoint3Triangle3.pdf

Comment (voferreira) [reply source/core/lethe_grid_tools.cc:1423]: Nice! Sorry for insisting.

Comment (voferreira) [reply source/core/lethe_grid_tools.cc:1420]: Could you add this to the comment as well?

**Comment by blaisb** [comment]
@acdaigneault if this is good for you I will merge after CI. I just need your approval.

**Pull request merged**
